vendors/st/boards/stm32l475_discovery/ports/common_io/iot_uart.c

```diff
@@ -0,0 +1,689 @@
+                        lError = IOT_UART_SUCCESS;
+                    }
+                }
+                else
+                {
+                    lError = IOT_UART_BUSY;
+                }
+
+                break;
+
+            case eUartGetConfig:
+
+                ( ( IotUARTConfig_t * ) pvBuffer )->ulBaudrate = pxHuart->Init.BaudRate;
+                ( ( IotUARTConfig_t * ) pvBuffer )->xParity = pxHuart->Init.Parity;
+                ( ( IotUARTConfig_t * ) pvBuffer )->ucWordlength = pxHuart->Init.WordLength;
+                ( ( IotUARTConfig_t * ) pvBuffer )->xStopbits = pxHuart->Init.StopBits;
+                ( ( IotUARTConfig_t * ) pvBuffer )->ucFlowControl = pxHuart->Init.HwFlowCtl;
+
+                lError = IOT_UART_SUCCESS;
+
+                break;
+
+            case eGetTxNoOfbytes:
+
+                *( int32_t * ) pvBuffer = ( int32_t ) ( pxHuart->TxXferSize - pxHuart->TxXferCount );
+                lError = IOT_UART_SUCCESS;
+
+                break;
+
+            case eGetRxNoOfbytes:
+
+                *( int32_t * ) pvBuffer = ( int32_t ) ( pxHuart->RxXferSize - pxHuart->RxXferCount );
+                lError = IOT_UART_SUCCESS;
+
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_cancel( IotUARTHandle_t const pxUartPeripheral )
+{
+    int32_t lError = IOT_UART_INVALID_VALUE;
+
+    if( ( pxUartPeripheral != NULL ) && ( pxUartPeripheral->sOpened == IOT_UART_OPENED ) )
+    {
+        /* lint !e9079 !e9087 handle passed should be of UART_HandleTypeDef *. */
+        UART_HandleTypeDef * pxHuart = ( UART_HandleTypeDef * ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->pxHuart;
+
+        if( HAL_UART_GetState( pxHuart ) == HAL_UART_STATE_READY )
+        {
+            lError = IOT_UART_NOTHING_TO_CANCEL;
+        }
+        else if( HAL_UART_Abort( pxHuart ) == HAL_OK )
+        {
+            lError = IOT_UART_SUCCESS;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+void HAL_UART_ErrorCallback( UART_HandleTypeDef * huart )
+{
+    /* Propogate Error message back to user.
+     * Does not use common-io standard API that signals if error was write or read,
+     * but this is not possible with ST API in Full-Duplex mode, as it is unkown if error came from Rx or Tx */
+    uint32_t error = HAL_UART_GetError( huart );
+
+    if( error != HAL_UART_ERROR_NONE )
+    {
+        uint32_t i = 0;
+
+        for( ; i < sizeof( xUartHandleMap ) / sizeof( UART_HandleTypeDef ); i++ )
+        {
+            if( huart->Instance == xUartHandleMap[ i ].Instance )
+            {
+                if( pxUarts[ i ]->xUartCallback != NULL )
+                {
+                    pxUarts[ i ]->xUartCallback( huart->ErrorCode, pxUarts[ i ]->pvUserCallbackContext );
+                }
+
```

Comment:
It's not necessary to give the semaphore back, as per to the two ISR callbacks below?

Reply:
I think I missed it. Fixed.